crates/op-consensus/src/receipt/envelope.rs

```diff
@@ -0,0 +1,216 @@
+use crate::{OpDepositReceipt, OpReceiptWithBloom, OpTxType};
+use alloy_eips::eip2718::{Decodable2718, Encodable2718};
+use alloy_primitives::{Bloom, Log};
+use alloy_rlp::{length_of_length, BufMut, Decodable, Encodable};
+
+/// Receipt envelope, as defined in [EIP-2718], modified for OP Stack chains.
+///
+/// This enum distinguishes between tagged and untagged legacy receipts, as the
+/// in-protocol merkle tree may commit to EITHER 0-prefixed or raw. Therefore
+/// we must ensure that encoding returns the precise byte-array that was
+/// decoded, preserving the presence or absence of the `TransactionType` flag.
+///
+/// Transaction receipt payloads are specified in their respective EIPs.
+///
+/// [EIP-2718]: https://eips.ethereum.org/EIPS/eip-2718
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "serde", serde(tag = "type"))]
+#[non_exhaustive]
+pub enum OpReceiptEnvelope<T = Log> {
+    /// Receipt envelope with no type flag.
+    #[cfg_attr(feature = "serde", serde(rename = "0x0", alias = "0x00"))]
+    Legacy(OpReceiptWithBloom<T>),
```

Comment:
hmm, this doesn't feel right to use the Opdepositreceipt for all variants

can we replicate the eth variants and only modify the deposit variant?